packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+                e.setHeader("Used test-only constant `{}` in non-test file", lit.symbol.show(ctx));
+            }
+        }
+
+        // no need to check visibility for these cases
+        auto otherPackage = ctx.state.packageDB().getPackageNameForFile(otherFile);
+        if (!otherPackage.exists() || this->package.mangledName() == otherPackage) {
+            return tree;
+        }
+
+        // Did we resolve a packaged symbol in unpackaged code?
+        if (!this->package.exists()) {
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::PackagedSymbolInUnpackagedContext)) {
+                e.setHeader("Packaged constant `{}` used in an unpackaged context", lit.symbol.show(ctx));
+            }
+            return tree;
+        }
+
+        auto importType = this->package.importsPackage(otherPackage);
+        if (!importType.has_value()) {
+            // We failed to import the package that defines the symbol
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::MissingImport)) {
+                auto &pkg = ctx.state.packageDB().getPackageInfo(otherPackage);
+                e.setHeader("No import provides `{}`", lit.symbol.show(ctx));
+                bool isTestImport = otherFile.data(ctx).isPackagedTest();
+                if (auto exp = this->package.addImport(ctx, pkg, isTestImport)) {
+                    e.addAutocorrect(std::move(exp.value()));
+                }
+                e.addErrorLine(pkg.declLoc(), "Defined here");
+            }
+
+            return tree;
+        } else if (*importType == core::packages::ImportType::Test && !this->insideTestFile) {
+            // We used a symbol from a `test_import` in a non-test context
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::UsedTestOnlyName)) {
+                e.setHeader("Used `{}` constant `{}` in non-test file", "test_import", lit.symbol.show(ctx));
+                auto &pkg = ctx.state.packageDB().getPackageInfo(otherPackage);
+                if (auto exp = this->package.addImport(ctx, pkg, false)) {
+                    e.addAutocorrect(std::move(exp.value()));
+                }
+                e.addErrorLine(pkg.declLoc(), "Defined here");
+            }
+        }
+
+        bool isExported = false;
+        if (lit.symbol.isClassOrModule()) {
+            isExported = lit.symbol.asClassOrModuleRef().data(ctx)->flags.isExported;
+        } else if (lit.symbol.isFieldOrStaticField()) {
+            isExported = lit.symbol.asFieldRef().data(ctx)->flags.isExported;
+        }
+
+        // Did we use a constant that wasn't exported?
+        if (!isExported) {
+            if (auto e = ctx.beginError(lit.loc, core::errors::Packager::UsedPackagePrivateName)) {
+                auto &pkg = ctx.state.packageDB().getPackageInfo(otherPackage);
+                e.setHeader("Package `{}` does not export `{}`", pkg.show(ctx), lit.symbol.show(ctx));
```

Comment:
```suggestion
                e.setHeader("`{}` exists but has not been exported by `{}", lit.symbol.show(ctx), pkg.show(ctx));
```

Reply:
I'm also happy to push this to a follow up change, as it looks like this error message was chosen for compatibility with what existed.